### ISSUE TITLE
Drop rule prohibiting passing exception instances to fail

### ISSUE THREAD

**Issue by gregnavis** [rubocop/ruby-style-guide#395]
# Problem

The style guide recommends against passing an exception instance to `fail`. From the comments to https://github.com/bbatsov/ruby-style-guide#exception-class-messages I conclude the reason is the supposed inability to specify a backtrace. This is not true as I demonstrate below. Please correct me if my reasoning is wrong. I tested and used the source of MRI 2.1.5.

In my projects I very often add contextual information to exceptions. For example

``` ruby
class UnknownUser < Exception
  attr_reader :user_name

  def initialize(user_name)
    @user_name = user_name
  end
end

# Somewhere in code ...
fail(UnknownUser.new('grn'))
```

Rubocop complains on the last line with:

```
test2.rb:10:1: C: Provide an exception class and message as arguments to fail.
fail(UnknownUser.new('grn'))
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

However `'grn'` in this case is _not an error message_ but a context of the exception. It could be an error code, IP address and port number that I tried to connect to, etc.
# Rationale

`fail(object, message, backtrace)` works like this:
1. it calls `object.exception(message)` which should return the exception to raise with the attached message; `Exception#exception` **clones** the exception, sets the message and resets the backtrace specified in the `Exception` instance
2. sets the backtrace of the exception returned by `object.exception` to the third argument of `fail`.

In the example `UnknownUser` exception it _is_ possible to define a custom backtrace:

``` ruby
fail(UnknownUser.new('grn'), 'user not found', my_backtrace)
```

See below for more details.
# Proof of Concept

Below is a script that uses `fail(exception_instance, message, backtrace)`.

``` ruby
class MyException < Exception
  attr_reader :error_code

  def initialize(error_code)
    @error_code = error_code
  end
end

def do_the_work(reset_backtrace)
  begin
    fail(MyException.new(1), 'My message', reset_backtrace ? [] : caller)
  rescue MyException => e
    puts "Rescued MyException with code #{e.error_code} and message #{e.message}"
    puts "Backtrace:"
    puts e.backtrace
  end
end

def test_exception(reset_backtrace)
  do_the_work(reset_backtrace)
end

test_exception(false)
puts
test_exception(true)
```

The script produces:

```
Rescued MyException with code 1 and message My message
Backtrace:
test.rb:20:in `test_exception'
test.rb:23:in `<main>'

Rescued MyException with code 1 and message My message
Backtrace:
```
# Details

`fail` is implemented in `rb_f_raise` which looks like:

``` c
static VALUE
rb_f_raise(int argc, VALUE *argv)
{
    VALUE err;
    if (argc == 0) {
        err = get_errinfo();
        if (!NIL_P(err)) {
            argc = 1;
            argv = &err;
        }
    }
    rb_raise_jump(rb_make_exception(argc, argv));

    UNREACHABLE;
}
```

The exception to raise is returned from `rb_make_exception` which delegates its work to `make_exception`. Its definition is (I added some comments):

``` c
static VALUE
make_exception(int argc, VALUE *argv, int isstr)
{
    VALUE mesg, exc;
    ID exception;
    int n;

    mesg = Qnil;
    switch (argc) {
      case 0:
        break;
      case 1:
        exc = argv[0];
        if (NIL_P(exc))
            break;
        if (isstr) {
            mesg = rb_check_string_type(exc);
            if (!NIL_P(mesg)) {
                mesg = rb_exc_new3(rb_eRuntimeError, mesg);
                break;
            }
        }
        n = 0;
        goto exception_call;

      case 2:
      case 3: // the code below handles fail(object, message[, backtrace])
        exc = argv[0];
        n = 1;
      exception_call:
        if (exc == sysstack_error) return exc;
        // call object.exception(message); this method MUST be defined
        CONST_ID(exception, "exception");
        mesg = rb_check_funcall(exc, exception, n, argv+1);
        if (mesg == Qundef) {
            rb_raise(rb_eTypeError, "exception class/object expected");
        }
        break;
      default:
        rb_check_arity(argc, 0, 3);
        break;
    }
    // The condition in case of fail with any arguments is true.
    if (argc > 0) {
        // object.exception(message) should return something of the kind of Exception
        if (!rb_obj_is_kind_of(mesg, rb_eException))
            rb_raise(rb_eTypeError, "exception object expected");
        // If backtrace was specified then attach it to the exception object.
        if (argc > 2)
            set_backtrace(mesg, argv[2]);
    }

    return mesg;
}
```

You can see that the backtrace is attached to the object whenever it is passed. The built-in `exception` method is implemented as:

``` c
/*
 *  Document-method: exception
 *
 *  call-seq:
 *     exc.exception(string)  ->  an_exception or exc
 *
 *  With no argument, or if the argument is the same as the receiver,
 *  return the receiver. Otherwise, create a new
 *  exception object of the same class as the receiver, but with a
 *  message equal to <code>string.to_str</code>.
 *
 */

static VALUE
exc_exception(int argc, VALUE *argv, VALUE self)
{
    VALUE exc;

    if (argc == 0) return self;
    if (argc == 1 && self == argv[0]) return self;
    exc = rb_obj_clone(self);
    exc_initialize(argc, argv, exc);

    return exc;
}

/*
 * call-seq:
 *    Exception.new(msg = nil)   ->  exception
 *
 *  Construct a new Exception object, optionally passing in
 *  a message.
 */

static VALUE
exc_initialize(int argc, VALUE *argv, VALUE exc)
{
    VALUE arg;

    rb_scan_args(argc, argv, "01", &arg);
    rb_iv_set(exc, "mesg", arg);
    rb_iv_set(exc, "bt", Qnil);

    return exc;
}
```

**Comment by wazery** [comment]
:+1:

**Comment by gregnavis** [comment]
Any thoughts on this? If we agree this is a good idea I can contribute a patch.

**Comment by bbatsov** [comment]
I'm certainly open to improving the wording of the rule.

**Comment by gregnavis** [comment]
@bbatsov it's not about rewording but rather dropping the rule altogether.

**Comment by mikegee** [comment]
I believe your proof that the justification comment is incorrect.

> Note that there is no way to do `raise SomeException.new('message'), backtrace`.

But I still want the style guide to prefer one way of calling `raise` & `fail` over the other. In general, the style guide should prefer one option over the other for every equivalent Ruby construct.

So I would support a change to the justification comments in the example code, but not a change to the recommendation.

**Comment by gregnavis** [comment]
@mikegee, thank you, that's a good point. I think we should then prefer:

> fail ExceptionClass.new("exception details")

over

> fail ExceptionClass, "exception message"

Would you agree?

**Comment by mikegee** [comment]
lol nope 😝 

That's exactly the opposite of what I mean. The style guide currently recommends:

> Prefer supplying an exception class and a message as two separate arguments to raise, instead of an exception instance.

It goes on to provide a bad justification for this recommendation. I just think we need to delete the bad justification.

**Comment by gregnavis** [comment]
Ah, I see. I'm arguing in favour of flipping that rule to require `fail ExceptionClass.new(...)`. A summary of my reasoning is this:

1. `fail ExceptionClass, "message"` has a drawback of not allowing passing parameters to the exception instance.
2. The only justification for it was that it makes it possible to specify a custom backtrace.
3. This justification is incorrect.
4. Therefore there's no reason to require this more limiting form.

**Comment by pirj** [comment]
Related:
 - https://github.com/rubocop-hq/ruby-style-guide/issues/395
 - https://github.com/rubocop-hq/ruby-style-guide/issues/747

It makes total sense to allow an exception to the guideline when a complementary contextual information is passed to the exception initializer.

The current wording is:

> Prefer supplying an exception class and a message as two separate arguments

And it's correct. It doesn't say anything about complementary objects passed to custom exception initializer. I see a space of imprevement to include such a note.


Speaking of RuboCop's `Style/RaiseArgs`, at the moment it's sophisticated enought to tell a built-in exception from a custom one. Wwith the default `exploded` style) for a custom exception is initialized with an argument **and** a message (https://github.com/rubocop-hq/rubocop/blob/8322f3aa371d668fde2ff133d592ff654a56ea01/spec/rubocop/cop/style/raise_args_spec.rb#L132):
```ruby
expect_no_offenses('raise Ex.new(entity), message')
```
Unfortunaly, there is no example of this in its specs (I've written the one above myself).

Also, even if you set the message internally in the custom exception and don't pass it to `raise`, `RaiseArgs` [can still ignore such custom exceptions given you've explicitly added them to `AllowedCompactTypes` setting](https://github.com/rubocop-hq/rubocop/blob/8322f3aa371d668fde2ff133d592ff654a56ea01/spec/rubocop/cop/style/raise_args_spec.rb#L271):

```ruby
let(:cop_config) do
  {
    'EnforcedStyle' => 'exploded',
     'AllowedCompactTypes' => ['Ex1']
  }
end
# ...
  expect_no_offenses('raise Ex1.new(msg)')
```

So, from my point of view, what needs to be done:
 - [ ] add a note to the guideline that it's acceptable to pass an instance of a custom exception that accepts an argument **followed** by a message (`raise MyEx.new('some text'), 'message'`)
 - [ ] add a note to the guideline that it's acceptable to pass an instance of a custom exception that accepts an non-message argument **without** a message argument (`raise MyEx.new(i_m_not_a_message)`)
 - [ ] add a spec for `expect_no_offenses('raise Ex.new(entity), message')` for `exploded` to `Style/RaiseArgs` cop's specs

Does this sound reasonable?

**Comment by gregnavis** [comment]
I think we're discussing two separate topics.

First, is the fact that the style guide still mentions:

> Note that there is no way to do `raise SomeException.new('message'), backtrace`.

I understand this statement as "If you pass an exception _instance_ then you can't specific a backtrace". If that's the intended meaning then it's false and the justification just be dropped. If that's not the indented meaning then I suggest rewording it a bit.

Second, there's the question what the default rule should be. I personally don't have a strong opinion on this one. I assumed the rule is there due to that backtrace confusion. If that's not the case and we want to keep the rule for consistency then I'm totally fine with that.

**Comment by pirj** [comment]
As far as I understand, the guideline says that if an instance of exception is passed to `raise`, it won't accept backtrace as the **second** argument. Surely, as your example shows, it is possible to pass the backtrace as the third one.

My feeling is that all the controversy around this guideline is concentrated in `"message"` argument to the custom exception.
The guideline seems to point out if one does not define an initializer for a custom exception, e.g.:
```ruby
class MyEx < Exception
end
```
that it will accept a message argument. And `raise` with an instance of this custom exception would correctly use this message if no other arguments are passed.

However, if you pass a second argument, message, the one you've passed to custom exception initializer will be lost:
```ruby
class MyEx < Exception
end

begin
  raise MyEx.new("LOST MESSAGE"), "passed message"
rescue MyEx => e
  puts e.message
end
```
outputs `passed message`.

I hope you'd agree with the statement that in this specific case it doesn't make sense to pass the message argument to the custom exception initializer, and it would better be passed as the second argument to `raise`. Doing so would still allow passing a custom backtrace, but at the cost of a `"LOST MESSAGE"`.

Undoubtedly, custom exceptions can be much more ...ergh... custom. And this guideline won't apply to them.

If you feel like you have a good wording for a new clarifying guideline in mind, please don't hesitate to send a pull request.
To me, this guideline should point out that if an instance of a custom exception that defines an initializer that accepts arguments that are not a message, it is ok to pass an instance of such a custom exception to `raise`.
It is also worth mentioning that in this case, the message should be still passed as the second argument to `raise`.
